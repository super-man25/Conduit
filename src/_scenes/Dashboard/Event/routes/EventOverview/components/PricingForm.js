// @flow
import React from 'react';
import {
  Flex,
  PrimaryButton,
  SecondaryButton,
  H4,
  Loader,
  NumberInputField
} from '_components';

import styled from 'styled-components';

const Row = styled.div`
  margin: 4px 0;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  min-height: 24px;
`;

const Label = styled.label`
  align-self: flex-start;
`;

const Text = styled.span`
  width: 60px;
  margin-left: 16px;
  display: inline-block;
  text-align: right;
  font-size: 1.125rem;
`;

const Input = styled.input`
  box-sizing: border-box;
  width: 60px;
  margin-left: 16px;
`;

const Line = styled.div`
  height: 1px;
  width: 100%;
  background-color: #aaa;
`;

const Group = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
`;

const ButtonGroup = styled.div`
  display: flex;
  padding-top: 20px;

  > *:not(:first-child) {
    margin-left: 8px;
  }
`;

type FormValues = {
  eventScore?: number,
  eventScoreModifier?: number,
  spring?: number,
  springModifier?: number
};

type Props = {
  initialValues: FormValues,
  onChange?: (formValues: FormValues) => any,
  onSubmit: (
    formValues: FormValues,
    onSuccess: () => void,
    onError: () => void
  ) => any
};

type State = {
  values: FormValues,
  initialValues: FormValues,
  editing: boolean,
  submitting: boolean
};

const SPRING_DECIMALS = 2;
const SCORE_DECIMALS = 3;

function safeAdd(a: any, b: any, fix: number) {
  const aVal = +a;
  const bVal = +b;

  if (isNaN(aVal) || isNaN(bVal)) {
    return '--';
  }

  return (aVal + bVal).toFixed(fix);
}

function fixedOrDash(number: any, fix: number) {
  const numberVal = +number;

  if (isNaN(numberVal)) {
    return '--';
  }

  return numberVal.toFixed(fix);
}

export class PricingForm extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      values: props.initialValues,
      initialValues: props.initialValues,
      editing: false,
      submitting: false
    };
  }

  handleChange = (field: string, value: any) => {
    const oldValues = this.state.values;
    const values = { ...oldValues, [field]: value };

    this.setState({ values });
    if (this.props.onChange) {
      this.props.onChange(values);
    }
  };

  handleCancel = () => {
    if (this.props.onChange) {
      this.props.onChange(this.state.initialValues);
    }
    this.setState({
      values: this.state.initialValues,
      editing: false
    });
  };

  handleEdit = () => {
    this.setState({ editing: true });
  };

  handleSuccess = () => {
    this.setState({
      submitting: false,
      editing: false,
      initialValues: this.state.values
    });
  };

  handleError = () => {
    this.setState({ submitting: false, editing: true });
  };

  handleSave = () => {
    this.setState({ submitting: true });
    this.props.onSubmit(
      this.state.values,
      this.handleSuccess,
      this.handleError
    );
  };

  render() {
    const { editing, submitting, values } = this.state;

    return (
      <>
        <Flex direction="row" justify="space-around">
          <Group>
            <H4
              style={{
                alignSelf: 'flex-start',
                padding: '0 4px',
                margin: '0',
                marginBottom: '12px'
              }}
            >
              Event Score
            </H4>
            <div style={{ maxWidth: '160px', paddingLeft: '20px' }}>
              <Row style={{ fontSize: '1.1rem' }}>
                <Label>Base:</Label>
                <Text>{fixedOrDash(values.eventScore, SCORE_DECIMALS)}</Text>
              </Row>
              <Row style={{ fontSize: '1.1rem' }}>
                <Label>Modifier:</Label>
                {editing ? (
                  <NumberInputField
                    component={Input}
                    value={values.eventScoreModifier}
                    onChange={(e) =>
                      this.handleChange('eventScoreModifier', e.target.value)
                    }
                  />
                ) : (
                  <Text>
                    {fixedOrDash(values.eventScoreModifier, SCORE_DECIMALS)}
                  </Text>
                )}
              </Row>
              <Line />
              <Row style={{ fontSize: '1.1rem', fontWeight: 600 }}>
                <Label>Final:</Label>
                <Text>
                  {safeAdd(
                    values.eventScore,
                    values.eventScoreModifier,
                    SCORE_DECIMALS
                  )}
                </Text>
              </Row>
            </div>
          </Group>
          <Group>
            <H4
              style={{
                alignSelf: 'flex-start',
                padding: '0 4px',
                margin: '0',
                marginBottom: '12px'
              }}
            >
              Spring Value
            </H4>
            <div style={{ maxWidth: '160px', paddingLeft: '20px' }}>
              <Row style={{ fontSize: '1.1rem' }}>
                <Label>Base:</Label>
                <Text>{fixedOrDash(values.spring, SPRING_DECIMALS)}</Text>
              </Row>
              <Row style={{ fontSize: '1.1rem' }}>
                <Label>Modifier:</Label>
                {editing ? (
                  <NumberInputField
                    component={Input}
                    value={values.springModifier}
                    onChange={(e) =>
                      this.handleChange('springModifier', e.target.value)
                    }
                  />
                ) : (
                  <Text>
                    {fixedOrDash(values.springModifier, SPRING_DECIMALS)}
                  </Text>
                )}
              </Row>
              <Line />
              <Row style={{ fontSize: '1.1rem', fontWeight: 600 }}>
                <Label>Final:</Label>
                <Text>
                  {safeAdd(
                    values.spring,
                    values.springModifier,
                    SPRING_DECIMALS
                  )}
                </Text>
              </Row>
            </div>
          </Group>
        </Flex>
        <ButtonGroup>
          {editing ? (
            <>
              <SecondaryButton onClick={this.handleCancel}>
                Cancel
              </SecondaryButton>
              <PrimaryButton disabled={submitting} onClick={this.handleSave}>
                {submitting ? <Loader small /> : 'Save'}
              </PrimaryButton>
            </>
          ) : (
            <SecondaryButton onClick={this.handleEdit}>Edit</SecondaryButton>
          )}
        </ButtonGroup>
      </>
    );
  }
}
