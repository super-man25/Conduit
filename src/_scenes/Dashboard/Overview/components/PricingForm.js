// @flow
import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { Box, Flex, H4, NumberInputField, Text } from '_components';
import { cssConstants } from '_constants';
import { fixedOrDash } from '_helpers/string-utils';
import { PendingFactors } from '_models';
import { PricingTableHeader } from './PricingTableHeader';
import {
  velocityScore,
  finalEventScore,
  finalSpringValue,
  SPRING_DECIMALS,
  SCORE_DECIMALS,
} from '_helpers/pricing-utils';

const StyledPricingForm = styled(Flex)`
  border-right: 1px solid ${cssConstants.PRIMARY_LIGHT_GRAY};
  padding: 25px;
  width: 100%;
`;

const EditButton = styled.div`
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  font-weight: bold;
  color: ${({ disabled }) =>
    disabled ? cssConstants.PRIMARY_LIGHT_GRAY : cssConstants.PRIMARY_BLUE};
`;

const Input = styled.input`
  width: 60%;
  margin-left: 16px;
  padding: 0.5rem 1rem;
  text-align: right;
  border-radius: 3px;
  border: 1px solid ${cssConstants.PRIMARY_GRAY};
`;

type Props = {
  pendingFactors: PendingFactors,
  handleChange: (event: SyntheticInputEvent<HTMLInputElement>) => void,
  eventId: number,
  fetchAutomatedSpring: (id: number, eventScore: ?number) => void,
  pricingError: ?Error,
  pricingPreview: {
    error: ?Error,
    loading: boolean,
  },
  isEditing: boolean,
  setIsEditing: Function,
};

export const PricingForm = (props: Props) => {
  const {
    handleChange,
    pendingFactors: {
      eventScore,
      eventScoreModifier,
      spring,
      springModifier,
      velocityFactor,
    },
    fetchAutomatedSpring,
    eventId,
    pricingError,
    pricingPreview: {
      error: pricingPreviewError,
      loading: pricingPreviewLoading,
    },
    isEditing,
    setIsEditing,
  } = props;

  const [timer, setTimer] = useState(null);

  // fires request only when user stops typing for 0.5 seconds to prevent over-firing
  const waitUntilFinished = () => {
    clearTimeout(timer);
    setTimer(setTimeout(updateSpring, 500));
  };

  const updateSpring = () => {
    const total = +finalEventScore(
      eventScore,
      velocityFactor,
      eventScoreModifier
    );
    fetchAutomatedSpring(eventId, total);
  };

  useEffect(() => (pricingError ? setIsEditing(true) : setIsEditing(false)), [
    pricingError,
    setIsEditing,
  ]);

  const editButtonDisabled =
    !eventScore || !!pricingPreviewError || pricingPreviewLoading;

  return (
    <StyledPricingForm direction="column">
      <Flex align="center" justify="space-between">
        <H4 margin="0">Modifiers</H4>
        {!isEditing && (
          <EditButton
            onClick={() => !editButtonDisabled && setIsEditing(true)}
            disabled={editButtonDisabled}
          >
            Edit
          </EditButton>
        )}
      </Flex>
      <Box padding="1.5rem">
        <PricingTableHeader headers={['Event Score', 'Spring Value']} />
        <Flex padding="14px 0" direction="row" minHeight="36px" align="center">
          <Box width="33%">
            <Text color={cssConstants.PRIMARY_GRAY}>Predicted</Text>
          </Box>
          <Box width="33%">
            <Text textAlign="right" className="private">
              {fixedOrDash(eventScore, SCORE_DECIMALS)}
            </Text>
          </Box>
          <Box width="33%">
            <Text textAlign="right" className="private">
              {fixedOrDash(spring, SPRING_DECIMALS)}
            </Text>
          </Box>
        </Flex>

        <Flex
          style={{
            borderTop: `1px solid ${cssConstants.PRIMARY_LIGHTER_GRAY}`,
            borderBottom: `1px solid ${cssConstants.PRIMARY_LIGHTER_GRAY}`,
          }}
          padding="14px 0"
          direction="row"
          align="center"
          minHeight="36px"
        >
          <Flex width="33%">
            <Text color={cssConstants.PRIMARY_GRAY}>Modifier</Text>
          </Flex>
          <Flex width="33%" justify="flex-end">
            {isEditing ? (
              <NumberInputField
                component={Input}
                name="eventScoreModifier"
                value={eventScoreModifier}
                onKeyUp={waitUntilFinished}
                onChange={handleChange}
              />
            ) : (
              <Text textAlign="right" className="private">
                {fixedOrDash(eventScoreModifier, SCORE_DECIMALS)}
              </Text>
            )}
          </Flex>
          <Flex width="33%" justify="flex-end">
            {isEditing ? (
              <NumberInputField
                component={Input}
                name="springModifier"
                value={springModifier}
                onChange={handleChange}
              />
            ) : (
              <Text textAlign="right" className="private">
                {fixedOrDash(springModifier, SPRING_DECIMALS)}
              </Text>
            )}
          </Flex>
        </Flex>

        <Flex
          style={{
            borderBottom: `1px solid ${cssConstants.PRIMARY_LIGHTER_GRAY}`,
          }}
          padding="14px 0"
          direction="row"
          align="center"
          minHeight="36px"
        >
          <Flex width="33%">
            <Text color={cssConstants.PRIMARY_GRAY}>Velocity</Text>
          </Flex>
          <Flex width="33%" justify="flex-end">
            <Text textAlign="right" className="private">
              {fixedOrDash(
                velocityScore(eventScore, velocityFactor),
                SCORE_DECIMALS
              )}
            </Text>
          </Flex>
          <Flex width="33%" justify="flex-end"></Flex>
        </Flex>

        <Flex padding="14px 0" direction="row" align="center" minHeight="36px">
          <Box width="33%">
            <Text weight="heavy">Final</Text>
          </Box>
          <Box width="33%">
            <Text weight="heavy" textAlign="right" className="private">
              {finalEventScore(eventScore, velocityFactor, eventScoreModifier)}
            </Text>
          </Box>
          <Box width="33%">
            <Text weight="heavy" textAlign="right" className="private">
              {finalSpringValue(spring, springModifier)}
            </Text>
          </Box>
        </Flex>
      </Box>
    </StyledPricingForm>
  );
};
