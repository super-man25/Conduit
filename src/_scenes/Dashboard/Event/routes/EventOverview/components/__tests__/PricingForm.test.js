import React from 'react';
import renderer from 'react-test-renderer';
import { PricingForm } from '../PricingForm';

const defaultProps = {
  initialValues: {
    eventScore: 1,
    eventScoreModifier: 1,
    spring: 1,
    springModifier: 1
  },
  pendingFactors: {
    eventScore: 1,
    eventScoreModifier: 1,
    spring: 1,
    springModifier: 1
  },
  onChange: jest.fn(),
  onSubmit: jest.fn(),
  eventId: 1,
  fetchAutomatedSpring: jest.fn(),
  onCancel: jest.fn(),
  pricingPreview: {
    error: null,
    loading: false
  }
};

it('render correctly', () => {
  const tree = renderer.create(<PricingForm {...defaultProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});
