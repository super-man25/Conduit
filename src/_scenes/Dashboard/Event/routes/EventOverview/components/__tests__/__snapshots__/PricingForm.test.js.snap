// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
Array [
  .c2 {
  border: none;
  min-width: 0;
  padding: 1rem 2rem;
  text-align: center;
  border-radius: 3px;
  font-size: 1rem;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  padding: 0;
  color: #024A8F;
  background-color: inherit;
}

.c2:disabled {
  cursor: not-allowed;
}

.c2:disabled {
  color: #D1D1D1;
  opacity: 0.5;
}

.c2:hover:not(:disabled) {
  color: #38A9DB;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.5rem 0;
}

.c1 {
  color: #070707;
  font-size: 18px;
  font-weight: normal;
  margin: 0;
}

<div
    className="c0"
  >
    <h4
      className="c1"
      color="#070707"
    >
      Modifiers
    </h4>
    <button
      className="c2"
      onClick={[Function]}
      style={
        Object {
          "fontWeight": 600,
        }
      }
    >
      Edit
    </button>
  </div>,
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 36px;
  padding: 14px 0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 33%;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 33%;
}

.c0 {
  padding: 1.5rem;
}

.c2 {
  width: 33%;
}

.c3 {
  font-size: 14px;
  color: #999999;
  margin: 0;
  text-align: right;
}

.c5 {
  font-size: undefinedpx;
  color: #999999;
  margin: 0;
}

.c6 {
  font-size: undefinedpx;
  color: #070707;
  margin: 0;
  text-align: right;
}

.c9 {
  font-size: undefinedpx;
  font-weight: 700;
  color: #070707;
  margin: 0;
}

.c10 {
  font-size: undefinedpx;
  font-weight: 700;
  color: #070707;
  margin: 0;
  text-align: right;
}

<div
    className="c0"
  >
    <div
      className="c1"
      direction="row"
    >
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c3"
          color="#999999"
          size={14}
        >
          EVENT SCORE
        </p>
      </div>
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c3"
          color="#999999"
          size={14}
        >
          SPRING VALUE
        </p>
      </div>
    </div>
    <div
      className="c4"
      direction="row"
    >
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c5"
          color="#999999"
        >
          PREDICTED
        </p>
      </div>
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c6"
        >
          1.00
        </p>
      </div>
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c6"
        >
          1.0000
        </p>
      </div>
    </div>
    <div
      className="c4"
      direction="row"
      style={
        Object {
          "borderBottom": "1px solid #EEEEEE",
          "borderTop": "1px solid #EEEEEE",
        }
      }
    >
      <div
        className="c7"
        width="33%"
      >
        <p
          className="c5"
          color="#999999"
        >
          MODIFIER
        </p>
      </div>
      <div
        className="c8"
        width="33%"
      >
        <p
          className="c6"
        >
          1.00
        </p>
      </div>
      <div
        className="c8"
        width="33%"
      >
        <p
          className="c6"
        >
          1.0000
        </p>
      </div>
    </div>
    <div
      className="c4"
      direction="row"
    >
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c9"
        >
          FINAL
        </p>
      </div>
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c10"
        >
          2.00
        </p>
      </div>
      <div
        className="c2"
        width="33%"
      >
        <p
          className="c10"
        >
          2.0000
        </p>
      </div>
    </div>
  </div>,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 > *:not(:first-child) {
  margin-left: 8px;
}

<div
    className="c0"
  />,
]
`;
